src/lib/factories.tsx

```diff
@@ -45,6 +52,34 @@ export function createShorthand(
   value?: ShorthandValue,
   options: CreateShorthandOptions = CREATE_SHORTHAND_DEFAULT_OPTIONS,
 ): React.ReactElement<Props> | null | undefined {
+  const valIsRenderFunction = typeof value === 'function' && !React.isValidElement(value)
+  if (valIsRenderFunction) {
+    const render: ShorthandRenderCallback = (shorthandValueOrRenderTree, renderTreeArg) => {
+      const shorthandArgType = {
+        isReactElement: React.isValidElement(shorthandValueOrRenderTree),
+        isRenderTreeFunc: typeof shorthandValueOrRenderTree === 'function',
+      }
+
+      const ShorthandElement = createShorthand(
```

Comment:
Would be a good idea to extract the rest of the logic in createShorthand in a separate method that we would called here, and maybe after this `if` statement is finished, just so that we don't make recursion here? If somebody for some reason change the condition inside the if statement we may introduce infinite recursion...

Reply:
I do understand your concerns here - let me try to play with that a bit, just the only criteria I am striving for is that final solution would be readable - and, actually, assured, it would be :) Thank you!

Reply:
have played with that - won't like to lose `const` in declaration (suggested option, unfortunately, will inevitably lead to `let` being used), but it seems that reasonable compromise is introduced :)